### DISCUSSION TITLE
Difference between WriterLoanExample and ZeroCopyExample

### DISCUSSION THREAD

**Discussion by duchengyao** [eProsima/Fast-DDS#2933] [Q&A] [answered]
I noticed two particularly similar examples `WriterLoanExample` and `ZeroCopyExample`, The difference is that one uses `take` and the other uses `take_next_sample()`.

Does it means `take_next_sample()` copy data, but `take()` wont' copy data?

**Comment by MiguelCompany** [comment, marked as answer]
> Does it means `take_next_sample()` copy data, but `take()` wont' copy data?

Correct, but incomplete.

`take_next_sample` will always copy data.
`take` will not copy data when called with empty `data_values` and `sample_infos`.

Check [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/dds_layer/subscriber/dataReader/readingData.html#loaning-and-returning-data-and-sampleinfo-sequences) and [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/subscriber/datareader.html#_CPPv4N8eprosima7fastdds3dds10DataReader4readER18LoanableCollectionR13SampleInfoSeq7int32_t15SampleStateMask13ViewStateMask17InstanceStateMask) for more information.

**Comment by duchengyao** [comment]
Thanks.
What happens if `data_values` or `sample_infos` is not empty, I didn't find the relevant instructions from the above two links.

Regards,

**Comment by MiguelCompany** [comment]
Literally copied from [here](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/subscriber/datareader.html#_CPPv4N8eprosima7fastdds3dds10DataReader4readER18LoanableCollectionR13SampleInfoSeq7int32_t15SampleStateMask13ViewStateMask17InstanceStateMask):

If the input `max_len == 0`, then the `data_values` and `sample_infos` collections will be filled with elements that are _loaned_ by the DataReader. On output, `owns` will be `false`, `len` will be set to the number of values returned, and `max_len` will be set to a value verifying `max_len >= len`. The use of this variant allows for zero-copy access to the data and the application will need to return the loan to the DataReader using the [return_loan](https://fast-dds.docs.eprosima.com/en/latest/fastdds/api_reference/dds_pim/subscriber/datareader.html#classeprosima_1_1fastdds_1_1dds_1_1DataReader_1a084b9a068abf972e27bfd201b6c6f15b) operation.

If input `max_len > 0` and the input `owns == true`, then the read operation will copy the Data values and SampleInfo values into the elements already inside the collections. On output, `owns` will be `true`, `len` will be set to the number of values copied, and `max_len` will remain unchanged. The use of this variant forces a copy but the application can control where the copy is placed and the application will not need to return the loan.

**Comment by duchengyao** [comment]
Thanks for your help ^_^
